### ISSUE TITLE
Question on the results in Figure 6

### ISSUE THREAD

**Issue by aktsonthalia** [fagp/sinkhorn-rebasin#11]
Hi, thanks for the great work! I had a question about the results in Figure 6 of the paper, where you show linear connectivity between VGG models trained on CIFAR-10. As far as I understand, you mention in the paper that [this code](https://worksheets.codalab.org/rest/bundles/0x5e77ad56e6c24487bbf832fac4d7ba9c/contents/blob/sinkhorn-rebasin/main_experiment1_cnn.py) was used to perform that experiment. So, are the models at the endpoints, exact permutations of each other in this case? Or is the shared code simply representative of the actual process (the endpoints are two different models trained from two different initializations, and different random seeds)?

Thanks in advance,
Ankit

**Comment by fagp** [comment]
Hi Ankit, thanks for your interest in our work. 

Your second statement is correct; the code is representative of the actual process. In this particular experiment, we used two endpoints that are completely different models trained from scratch to analyze how low the barrier we obtained could be with non-correlated models. The same goes for the ResNet experiment ([Figure 4 in the Supp. Mat,](https://openaccess.thecvf.com/content/CVPR2023/supplemental/Pena_Re-Basin_via_Implicit_CVPR_2023_supplemental.pdf) of the [camera-ready version](https://openaccess.thecvf.com/content/CVPR2023/papers/Pena_Re-Basin_via_Implicit_Sinkhorn_Differentiation_CVPR_2023_paper.pdf)).

Best regards,
Fidel

**Comment by aktsonthalia** [comment]
Thanks Fidel!

**Comment by aktsonthalia** [comment]
@fagp I have two additional questions if it's not too much trouble:

1. Do you use batch normalization in these models? "Batch normalization is supported for re-basin but not for LMC.": from this, I assume you don't (for Fig. 6 results)?
2. Would it be possible for you to provide the pretrained models for VGG11 and/or VGG19 on CIFAR-10? I have been trying to reproduce the result with two of my own models, but so far I haven't been successful. Having access to your models would help me with debugging my setup.

**Comment by heitorrapela** [comment]
Hello, I think @fagp can answer better, and maybe correct me. But here are my considerations:

1 -  If you see in the example code provided in this repo, for the Re-basin task, @fagp solves this by adjusting the statistics with the function "update_batchnorm" in the scripts. For LMC, as you are learning the curve after the rebasin, you have to consider updating the batch norm of the final model (because the final model did not see the data statistics); for that, you have to change the BNs using REPAIR, which is an extended idea that came after you combine models on weight space, you have to pass data to update the statistics of the final model (https://arxiv.org/pdf/2211.08403.pdf), you can check the open commit that we have in this repo, maybe can give you some guidance for this: https://github.com/fagp/sinkhorn-rebasin/commit/3d273afeae40eb9cb17ce49916375186b4296ab6, similar idea of updating the BNs are also used by SWA community, and you can check the following function: torch.optim.swa_utils.update_bn (https://github.com/pytorch/pytorch/blob/main/torch/optim/swa_utils.py#L226)

2 -  Here are the VGG11 and VGG19 models trained on CIFAR10 (https://huggingface.co/heitorrapela/vggs_cifar10/tree/main). Code used for training the models: https://github.com/chengyangfu/pytorch-vgg-cifar10). After you download, there are 20 VGGs 11 and 20 VGGs 19, with performance above 90% acc

**Comment by aktsonthalia** [comment]
Great, thanks a lot, @heitorrapela ! Yes, I'm aware of the batch normalization issue, good to know your code takes care of it too!

I'm unfortunately still unable to reproduce the results. I am using [this script](https://github.com/aktsonthalia/sinkhorn-rebasin/blob/main/examples/align_models_starlight.py), largely adapted from [yours](https://github.com/fagp/sinkhorn-rebasin/blob/main/examples/main_lmc_cnn.py), the only change I made was loading the model checkpoints. Am I doing something wrong? Are the [hyperparameters](https://github.com/aktsonthalia/sinkhorn-rebasin/blob/06a41db889452baf8eebbce4e48e9974406bbae4/examples/align_models_starlight.py#L63) wrong (I did try a variety of them, but the current ones I noted from the paper)?

I'm attaching the plot as well.

<img width="649" alt="Screenshot 2024-03-28 at 17 04 25" src="https://github.com/fagp/sinkhorn-rebasin/assets/53793862/86afc199-2cb2-47cb-b306-b25b55d5dd37">

Looking forward to hearing from you!

**Comment by heitorrapela** [comment]
Hello @aktsonthalia, as I told before, we solved the batch norm for the rebasin net, but not for the LMC in this script. Maybe change the rebasin to RndLoss can be better (because it is data driven), but you also need to change the hparams for that and for the LMC part as well. For LMC, if you dont correct the BNs, it is hard to decrease the error gap of the curve.

**Comment by aktsonthalia** [comment]
@heitorrapela thanks again for sharing the insight. I understand that batchnorm does not work with the current script. I will try to fix it using the resources you mentioned. Like you suggested, I will try the RndLoss experiment, and will try to tune hyperparameters as well. Thanks for suggesting.

 - However, the script that I linked, uses VGG without batchnorm (I checked the model and can confirm this), which is why I'm confused. 

- The data-free VGG-CIFAR10 experiment is quite important for my project, and if possible I would appreciate having access to the exact script used for this experiment. 

Best, Ankit

**Comment by heitorrapela** [comment]
@aktsonthalia, I have also added the BN models (https://huggingface.co/heitorrapela/vggs_cifar10/tree/main). It is known that the models with BNs, and also ResNet, tend to generate a loss landscape that is easier to optimize and connect. So, you are going to have, by default, a lower barrier when you generate the LMC, even without much tuning, compared with the models without BNs.

btw, for more details about that @fagp can answer better than me, but let us know if everything worked fine,

Best Regards,

**Comment by heitorrapela** [comment]
@aktsonthalia, I run here your script with some modifications on the hparams
VGG11bn - DistL2Loss, adamw lr lr=10.0
CKPT1 = os.path.join(CKPT_DIR, "VGG11bn_cifar10_0.924.pth")
CKPT2 = os.path.join(CKPT_DIR, "VGG11bn_cifar10_0.9213.pth")
you also have to change here modelA and modelB to VGG11bn: modelA = VGG("VGG11bn", in_channels=3, out_features=10, h_in=32, w_in=32)

![image](https://github.com/fagp/sinkhorn-rebasin/assets/14017613/76033bef-40e8-4c06-a126-1900a78c61fd)

**Comment by aktsonthalia** [comment]
@heitorrapela thanks a lot for the efforts, this looks great indeed! 

Unfortunately I didn't get the time to try anything myself yet because of some urgent commitments, but I will take a look again in the coming week. Cheers!

**Comment by fagp** [comment]
Hi guys, sorry for the delay in getting back to you. @aktsonthalia, hope you managed to reproduce those results alright. Big props to @heitorrapela for answering your questions and giving out advice on hyperparameters! I don't think there is anything else for me to answer. In any case, if you're still having issues with reproducing these results, don't hesitate to reach out. We would love to help you.
